kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,381 @@
+    :return: dict of type keys and count values
+    """
+    return {c: len(graph.nodes(to_node=c)) for c in reversed(HIERARCHY.nodes())}
+
+
+def _get_edge_counts(graph):
+    hnodes = set(HIERARCHY.nodes())
+    return len([e for e in graph.edges() if e[1] not in hnodes])
+
+
+def _validate_graph_relationships(graph, dict_of_dataframes):
+    """Perform tests on the graph to validate which of the hierarchy
+    rules have been broken by the study data.
+
+    :param graph: a graph representation of the data
+    :param dict_of_dataframes: dict with filename keys and dataframe values
+    :yield: test result dicts
+    """
+    assert isinstance(graph, Graph)
+    assert len(graph.nodes()) > 0
+
+    logger.info("Validating graph relationships...")
+
+    reverse_graph = Graph()
+    for e in graph.edges():
+        reverse_graph.add_edge(e[1], e[0])
+
+    membership_lookup = {
+        f: {col: set(df[col].values) for col in df.columns}
+        for f, df in dict_of_dataframes.items()
+    }
+
+    def find_in_files(node):
+        """Find which files a given node came from.
+
+        :param node: a node in the graph representing a data value
+        :return: a list of files
+        """
+        return [
+            f
+            for f, ml in membership_lookup.items()
+            if (node[0] in ml) and (node[1] in ml[node[0]])
+        ]
+
+    def cardinality(typeA, typeB, relation):
+        """Tests cardinality of connections between typeA and typeB.
+
+        :param typeA: a type node (nodes are inserted in the graph for value
+            types as well as values)
+        :param typeB: another type node
+        :param relation: one of the tuples from relations.RELATIONS
+        :return: a test result dict
+        """
+        desc, func = relation
+        logger.info(f"Testing {typeA} links to {desc} {typeB}...")
+        errors = []
+        A_nodes = graph.nodes(to_node=typeA)
+        if typeB in HIERARCHY.nodes(typeA):
+            g = graph
+        else:
+            g = reverse_graph
+        for n in A_nodes:
+            links = [c for c in g.nodes(n) if c[0] == typeB]
+            if not func(len(links)):
+                locs = {m: find_in_files(m) for m in ([n] + links)}
+                errors.append({"from": n, "to": links, "locations": locs})
+
+        desc = f"Each {typeA} links to {desc} {typeB}"
+        return _format_result("relationship", desc, bool(A_nodes), errors)
+
+    def gaps():
+        """Tests, according to the relationship hierarchy, that A always -> B
+        always -> C, and not A -> C without a B.
+
+        :return: a test result dict
+        """
+        logger.info("Looking for gaps in the graph hierarchy...")
+        errors = []
+        for n in graph.nodes():
+            links = [
+                m
+                for m in graph.nodes(from_node=n)
+                if (n[0] != m) and (m[0] not in HIERARCHY.nodes(from_node=n[0]))
```

Comment:
I'm not sure I understand this line. Should it be `n[0] != m[0]` instead?

Reply:
Nope! It's a test (no longer needed now) to make sure that we're not walking into the type node. Every node is two parts (type, value) except for the type nodes which are just type